packages/data/src/plugins/test/reducer.ts

```diff
@@ -7,32 +7,36 @@
  */
 import reducer from '../reducer';
 import TYPES from '../action-types';
+import { PluginsState } from '../types';
+import { Actions } from '../actions';
 
 const defaultState = {
 	active: [],
 	installed: [],
 	requesting: {},
 	errors: {},
 	jetpackConnectUrls: {},
-};
+	recommended: {},
+} as PluginsState;
 
 describe( 'plugins reducer', () => {
 	it( 'should return a default state', () => {
-		const state = reducer( undefined, {} );
+		const state = reducer( undefined, {} as Actions );
```

Comment:
Do you have any suggestions here? as the issue is the `test` or I should remove the de-structuring in the reducer. 
I could make all the items in `Actions` optional, but then I will have to handle all the optional cases within the switch statement, which doesn't quite make sense.
The ideal would be, that TS could determine the action key/values by the `type` value, but I can't find a good way to do that without writing specific types for each action (I was hoping to avoid that and be able to use the ReturnType).

So I be happy with an alternate solution?

Reply:
@louwie17 yep for sure. If `{}` is a legal argument then you simply union it with Action in the signature of the reducer, then you can use the switch statement to do type narrowing and you don't need to declare anything as optional.

Code example:

```js
type MyCoolAction = {
    type: 'cool-action';
    baz: string
};

type MyOtherCoolAction = {
    type: 'my-other-one';
    foo: number
}

type Action = MyCoolAction | MyOtherCoolAction


const reducer = (action: Action | {}) => {
    if ("type" in action) {
        switch (action.type) {
            case 'cool-action':
                console.log(action.baz);
                return {};
            case 'my-other-one': {
                // error
                console.log(action.baz);
                // ok
                console.log(action.foo);
                return {};
            }   
        }
    } else {
        return {};
    }
}
```

Here's the TS playground link if you want to have a play with it: https://www.typescriptlang.org/play?ssl=31&ssc=2&pln=1&pc=1#code/C4TwDgpgBAsiDCB7RAbAggY2AS0QOygF4oBvAKCkqlEgC4oByDZFAWgEMtc8GBuCqgCN2AL3oBnYACdseAOZkAvvzI1ocAPLAAFhClJUmHPiKkBlNfQYBbEK0Q699vBD7moAM2T08AV2uCekpkquDQRtymcAboXCYAPrAgWrr6LBH4IWTMeJJQUhAAJr4YeqYAFJzGePQZBIkkigCURAB8ZlRQ2B5Q5QBEan1dBFXcLeSdneIA7tjAGNq9o-gAdGrj7pNUGOzi0EwsHHE8tJtbkzniqBArKIhylccrwiJN-Ocf+RDAvlIEje9PpQdntGLZ7I4pM5XPQJkDJgB6BFQPRSRBSM5Ay7XW73R7VZ6iN6Yz5IqCIADWJI+2JQNzuD2WeBWXkQxPhkwKPz+pGU1KoiioJMU7kFEBQoLhWy5v3+fM6IpFZCAA

Reply:
Thanks @samueljseay :) that did the trick (almost) haha I did have to make sure I add a return type to each action specifying the specific action enum it used. Otherwise type stayed a generic `ACTION_TYPES` and the attributes wouldn't be passed down into the switch statement.
Much nicer this way though.

Reply:
Fixed here: https://github.com/woocommerce/woocommerce-admin/pull/6760/commits/7bddb6428d44ba15d343e9b52b269ecb36aa2b6e